main.cpp

```diff
@@ -402,3 +403,157 @@ int main(int argc, char** argv) {
+            printf("Failed to commit patches - Pretendo wave in place\n");
+            while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_RESTORE_FAIL); PresentMenu(); }
+            return -1;
+        }
+    }
+    RenderMenuLoading(task_percent(task++), "Verifying final patches... (nn_olv.rpl)");
+    PresentMenu();
+
+    {
+        std::ifstream is(nn_olv_path, std::ios::binary);
+        auto hash = rpx_hash(is);
+        if (hash.id == CURRENT_PRETENDO_NN_OLV) {
+            //we've still have the PN OLV
+            printf("Failed to commit patches - nn_olv corrupt!\n");
+            while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_RESTORE_FAIL); PresentMenu(); }
+            return -1;
+        }
+    }
+   RenderMenuLoading(task_percent(task++), "Verifying final patches... (NSSL cert)");
+    PresentMenu();
+
+    {
+        std::ifstream is(cert_path, std::ios::binary);
+        auto hash = cert_hash(is);
+        if (hash.id != CERT_ID_THWATE_PREMIUM_SERVER) {
+            //we've already modified wave, so this is always dangerous
+            printf("Failed to commit patches - nssl cert corrupt!\n");
+            //oh shit, this could brick - restore backup
+            bret = fast_copy_file(cert_bak_path, cert_path);
+            if (!bret) {
+                printf("Reverting nssl patches failed!\n");
+                while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_PATCH_FAIL_CERT_BRICK); PresentMenu(); }
+                return -1;
+            }
+
+            std::ifstream is2(cert_path, std::ios::binary);
+            auto hash2 = cert_hash(is2);
+            if (hash2.patch != CERT_PATCH_STATE_STOCK) {
+                printf("Reverting nssl patches failed!\n");
+                while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_PATCH_FAIL_CERT_BRICK); PresentMenu(); }
+                return -1;
+            } else if (hash2.id != CURRENT_SACRIFICIAL_CERT) {
+                printf("Reverting NSSL cert resulted in the wrong cert?\n");
+                while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_PATCH_FAIL_DANGEROUS); PresentMenu(); }
+                return -1;
+            }
+
+            while (WHBProcIsRunning()) { RenderMenuDone(MENU_DONE_PATCH_FAIL_DANGEROUS); PresentMenu(); }
+            return -1;
+        }
+        
+    }
+    //Flush Volumes
+    RenderMenuLoading(task_percent(19), "Flushing volumes...");
+    PresentMenu();
+
+    ret = FSAFlushVolume(fsaHandle, "/vol/storage_mlc01");
+    printf("FSAFlushVolume: %08x\n", ret);
+    //woo!
+    while (true) {
+        VPADStatus vpad;
+        VPADReadError error;
+        VPADRead(VPAD_CHAN_0, &vpad, 1, &error);
+        if (error == VPAD_READ_SUCCESS) {
+            if (vpad.trigger & VPAD_BUTTON_A) break;
+        }
+        RenderMenuDone(MENU_DONE_RESTORE_DONE);
+        PresentMenu();
+    }
+
+    printf("shutting down\n");
+
+    OSShutdown();
+    while (WHBProcIsRunning()) {}
+
+    return 0;
+}
```

Comment:
This whole thing should be indented by an extra 4 spaces - if you have a decent text editor this should be easy enough to do, if not let me know and I'll do a formatting pass after you